### PR TITLE
LibCompress+Utilities: Add LZMA

### PR DESCRIPTION
**Pull request by timschumi**
Pain.

This is annoyingly close to the reference implementation, at the very least the math-heavy parts that actually get into decoding instead of the high-level logic (granted, not like one can diverge much anyways if that's the only real source of "that is the formula you need to use" and "you need that many different probability counters and use them in a specific order"). The official specification is unfortunately basically just the reference implementation with a bit of text around the logic parts, and I wasn't able to find any human-language specifications that have actually been completed.

I hope that this is still acceptable, considering that the specification/reference implementation has been placed in the public domain.

## Remaining TODOs/discussion points

 - ~~Fuzzer: Not viable at the moment, since the archive can legitimately request up to 4GB of lookback buffer size, which instantly OOMs LibFuzzer.~~
 - ~~Specification quotes: The implementation currently features a mix of quotes from the specification and some comments of my own. Suggestions on how to make those visually distinct would be very welcome, if that is a priority.~~
 - ~~Tests: Will add tomorrow.~~
 - ~~Rough edges: Nothing is explicitly unimplemented at the moment, but there are some noticeable edge cases that I haven't looked into yet (for example, some of the testing files from the SDK don't even load properly).~~

### CODE REVIEW

File: Userland/Libraries/LibCompress/Lzma.h
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCompress/Lzma.h:29]: Should this not take `Handle<Stream>`?

File: Userland/Libraries/LibArchive/Lzma.h
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibArchive/Lzma.h:35]: Same here with Handle

File: Userland/Utilities/lzcat.cpp
Comment (kleinesfilmroellchen) [inline Userland/Utilities/lzcat.cpp:17]: Should probably have a short help text like "Decompress and print an LZMA archive"

File: Userland/Utilities/lzcat.cpp
Comment (kleinesfilmroellchen) [inline Userland/Utilities/lzcat.cpp:24]: Use PAGE_SIZE?

File: Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp
Comment (kleinesfilmroellchen) [inline Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp:14]: Can we fuzz all options *except* the dictionary size? i.e. overwrite the dictionary size option after reading the options from the provided data.

File: Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp
Comment (kleinesfilmroellchen) [inline Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp:29]: PAGE_SIZE as well probably

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCompress/Lzma.h
Comment (timschumi) [reply Userland/Libraries/LibCompress/Lzma.h:29]: It's called `MaybeOwned` now, but yes, it should (at the very least for LZMA2). :^)

File: Userland/Utilities/lzcat.cpp
Comment (timschumi) [reply Userland/Utilities/lzcat.cpp:24]: `PAGE_SIZE` is not consteval on Lagom, so we can't use it here. It also isn't nearly as important to align to pages compared to heap allocations. In this case, it really is just a nice and round number.

File: Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp
Comment (timschumi) [reply Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp:14]: We could do that, but that would result in files that look like LZMA archives (not that they have some kind of magic), but would have to be changed first to yield the same results as the fuzzer.

What we _could_ do is to define some arbitrary "largest reasonable dictionary size" and reject any test inputs that have a larger dictionary size than that. LibFuzzer has a dedicated return value to indicate "don't add this to the test corpus", which would be perfect for this.

Comment (kleinesfilmroellchen) [reply Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp]: That sounds like the ideal solution to me then.

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by timschumi** [comment]
Renamed `Archive::Lzma` to `Archive::LzmaDecompressor`, to match `Compress::LzmaDecompressor`. But now this should really be ready for merging. :^)

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by GMTA** [comment]
> Specification quotes: The implementation currently features a mix of quotes from the specification and some comments of my own. Suggestions on how to make those visually distinct would be very welcome, if that is a priority.

I try to explicitly reference section numbers when possible, and literally quote (`"The spec says that..."`) the parts that I copy over from the spec.

**Comment by timschumi** [comment]
I'm afraid the specification is a bit short on section numbers, and it uses section names/headers _very_ sparingly. I will try adding quotes around the quotes and prefixing them with some kind of distinctive text to see how it looks in practice.

**Comment by timschumi** [comment]
Changes in latest push:
 - Rebased onto and adapted to the latest Stream function name changes.
 - Removed some unnecessary braces around one line for-loops.
 - Visually marked verbatim specification comments by wrapping them in `"`.
 - Replaced the specification vs. reference implementation FIXME with a "Note", since I experimentally confirmed that the reference implementation is most likely correct.

**Review by kleinesfilmroellchen — APPROVED** [review]

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:139]: no need for else after return

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:190]: Not sure I understand this verify, if `read_with_seekback` didn't return an error `read_bytes` is a guarenteed to return 1 byte right? Otherwise it's a bug in the `read_with_seekback` implementation? If so, this seems a bit excessive, we should either assert this inside the implementation itself, or not at all.

Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.cpp:190]: (same for everywhere else below)

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:200]: We use this `0x300` constant in multiple places, maybe make it a constexpr constant?

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:242]: ditto regarding the helper from above

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:590]: nitpick, unnecessary curlies

File: Userland/Libraries/LibArchive/Lzma.cpp
Comment (ADKaster) [inline Userland/Libraries/LibArchive/Lzma.cpp:60]: Just to confirm: These properties on the bits are mandated by either the math here in this function or by the caller of this function? (i.e. we won't see any fuzzer crashes hitting these asserts, right?)

File: Userland/Libraries/LibArchive/Lzma.h
Comment (ADKaster) [inline Userland/Libraries/LibArchive/Lzma.h:37]: Archive::LzmaDecompressor::create() returns a Compress::LzmaDecompressor? This seems kind of odd, but then I didn't cross reference it with the other Decompressors that mostly just redirect to LibCompress.

File: Userland/Utilities/lzcat.cpp
Comment (ADKaster) [inline Userland/Utilities/lzcat.cpp:21]: Thoughts on using that "open file or standard stream" thing here? or does Posix/Linux lzcat not support `-` as an arg.

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:379]: nitpick, but `-=`

**Review by ADKaster — COMMENTED** [review]
Couple thoughts for clarification, but looks good :^)

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:446]: we already check this above, so this is unrechable. It would only be reachable if we did another iteration of the loop to check for the EOF marker, which we don't.

Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.cpp:446]: (I would keep the spec text, but add a comment explaining it's not unecessary, assuming you agree with me of course :slightly_smiling_face:)

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:482]: We set `m_found_end_of_stream_marker` (this is the only place where we do so), and immediately continue the loop, at which point we break out of it using the loop end condition at the top. We then go into the post-loop check for the code validity if we reached the end of stream marker.
AKA: This check for the final code value is duplicated, we should remove one of the copies.

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:488]: same as above

**Review by IdanHo — COMMENTED** [review]
nice work, mostly nitpicks

File: Userland/Libraries/LibArchive/Lzma.h
Comment (timschumi) [reply Userland/Libraries/LibArchive/Lzma.h:37]: The reason why I made this separate is because LZMA technically defines an 'archive' format (which is more of a stream container, really).

It is a really puny archive format, because all it does is prefixing it with a small header that notes down the decoder settings for the contained stream, but it is distinct from the other LibCompress-supported algorithms that really only dump the raw stream into a file and call it a day.

The reason why it returns a `Compress::LzmaDecompressor` is because we read the header upon creation of the stream, configure the underlying `LibCompress` stream once, and would then just delegate all the time anyways. Returning the underlying Stream directly and not implementing the whole Stream interface for Archive::LzmaDecompressor again seemed like a better idea.

Would we rather want to have a `Compress::LzmaDecompressor::create_from_archive` (/container?) and `Compress::LzmaDecompressor::create_from_raw_stream`, where one relies on the other?

File: Userland/Libraries/LibArchive/Lzma.h
Comment (timschumi) [reply Userland/Libraries/LibArchive/Lzma.h:38]: The question regarding the return type I have started to discuss in a response to Andrew, so we should probably continue to do that there exclusively.

File: Userland/Libraries/LibArchive/Lzma.h
Comment (timschumi) [reply Userland/Libraries/LibArchive/Lzma.h:37]: From Idans comment about the same topic:

> The distiniction between a compression format and an archival format, is that a compression format takes a stream of bytes and compresses/decompresses it, while an archive format packs together one or more files, with the file contents possibly being compressed using a compression format.

Ok, in that case LZMA 'archives' wouldn't be an actual archive because (in comparison to an LZMA compressed stream) they only provide the settings for the singular stream that they contain.

However, looking at what's to come, are LZMA2 and XZ archive formats then? Should they also rather be in LibCompress?

Neither of them have a concept of one or multiple files, both just encode a stream of data, but this time in multiple ways. LZMA2 splits the stream in parts and either stores them LZMA-compressed or uncompressed. XZ allows compressing its data stream using LZMA2 and/or a bunch of of other so-called filters. None of them know what 'files' are, if a user needs that they have to TAR it up first.

File: Tests/LibCompress/TestLzma.cpp
Comment (timschumi) [reply Tests/LibCompress/TestLzma.cpp:213]: > Are we going to fix this FIXME?

I plan to, yes. However, since my first attempt at this led to a bunch of erroneous failures and end-of-stream detection will have a significant effect on formats that embed LZMA, I decided to let it be for now and revisit it later, focusing on getting the main parts of the decoder (and the followup support for LZMA2 and XZ) finished up and merged first.

> Also, we might want to expose a is_corrupted helper that returns true if we reached the 'known' uncompressed size but m_range_decoder_code is non-0? The spec says that's a good indicator for an incorrectly decoded stream (and you already check that when we reach the explicit end indicator), so we can assert that returns true here.

The specification indeed mentions that this is a good way of checking for corruption, but it unfortunately doesn't seem to hold up if the archive contains an uncompressed size _and_ an end of stream marker (since it still has to read the end marker at that point). This is also what I suspect was causing my failures that shouldn't have been there.

In any case, the current mode of operation of this implementation is to stop and report an error even for things that would otherwise just be "corrupted", so I'd rather have it report that immediately if it even just notices that something is wrong.

**Review by timschumi — COMMENTED** [review]
Fixed a large part of the nitpicks, left comments on the others.

File: Userland/Libraries/LibArchive/Lzma.h
Comment (IdanHo) [reply Userland/Libraries/LibArchive/Lzma.h:37]: Comparing to the Gzip, which we already have implemented in serenity, it seems to match up: It's based on Deflate, it adds a short header at the start with the compression options, and is possibly used to compress tar archives.
Currently the implementation for Gzip is in LibCompress, matching my expectations at least.

The only unfortunate part about LZMA is that the name is used both for the raw stream and the stream with the header part.

I would say all 3 of LZMA, LZMA2 and XZ should end up in LibCompress.

File: Tests/LibCompress/TestLzma.cpp
Comment (IdanHo) [reply Tests/LibCompress/TestLzma.cpp:213]: I see, let's leave that to a future PR then :+1:

File: Userland/Libraries/LibCompress/Lzma.h
Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.h:83]: Ah, if it's just the chosen value per tree then it's fine, I thought `1 << 3` was being duplicated, but it sounds like it's actually 2 different values, low tree length, and medium tree length, which just ended up as the same value.

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.cpp:96]: Yep, I prefer so, simply to make the code a tiny bit more readable.

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.cpp:124]: nitpicks can be ignored when reasonable :slightly_smiling_face:

File: Userland/Libraries/LibCompress/Lzma.cpp
Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.cpp:185]: I think `probability_table_size` is fine

**Review by IdanHo — APPROVED** [review]
LGTM, thanks! Support for new compression formats is always exciting :^)

**Pull request merged**
